File: Day-4/CodeFOrces/contest/F.py
def var_index(v, pol):
    return 2 * v + (0 if pol else 1)

def solve_2sat():
    import sys
    input = sys.stdin.readline
    sys.setrecursionlimit(1000000)

    T = int(input())
    for _ in range(T):
        n, k = map(int, input().split())
        N = 2 * n
        g = [[] for _ in range(N)]
        gr = [[] for _ in range(N)]

        for _ in range(k):
            m = int(input())
            adj = [0] * n  # Represent bitset as integer for up to 64 nodes

            for _ in range(m):
                u, v = map(int, input().split())
                u -= 1
                v -= 1
                adj[u] |= 1 << v
                adj[v] |= 1 << u

            for u in range(n):
                for v in range(u + 1, n):
                    bu = adj[u] & ~(1 << u | 1 << v)
                    bv = adj[v] & ~(1 << u | 1 << v)
                    if bu != bv:
                        continue

                    has_edge = (adj[u] >> v) & 1
                    if has_edge:
                        # ¬C_u ∨ ¬C_v
                        a1 = var_index(u, True)
                        b1 = var_index(v, False)
                        g[a1].append(b1)
                        gr[b1].append(a1)

                        a2 = var_index(v, True)
                        b2 = var_index(u, False)
                        g[a2].append(b2)
                        gr[b2].append(a2)
                    else:
                        # C_u ∨ C_v
                        a1 = var_index(u, False)
                        b1 = var_index(v, True)
                        g[a1].append(b1)
                        gr[b1].append(a1)

                        a2 = var_index(v, False)
                        b2 = var_index(u, True)
                        g[a2].append(b2)
                        gr[b2].append(a2)

        used = [False] * N
        order = []

        def dfs1(u):
            used[u] = True
            for v in g[u]:
                if not used[v]:
                    dfs1(v)
            order.append(u)

        for i in range(N):
            if not used[i]:
                dfs1(i)

        comp = [-1] * N
        cid = 0

        def dfs2(u):
            comp[u] = cid
            for v in gr[u]:
                if comp[v] == -1:
                    dfs2(v)

        for u in reversed(order):
            if comp[u] == -1:
                dfs2(u)
                cid += 1

        ok = True
        for v in range(n):
            if comp[var_index(v, True)] == comp[var_index(v, False)]:
                ok = False
                break

        print("Yes" if ok else "No")
if __name__ == "__main__":
    solve_2sat()

File: Day-4/CodeFOrces/contest/F.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

const MAXN = 300

func varIndex(v int, pol bool) int {
	if pol {
		return 2 * v
	}
	return 2*v + 1
}

func solve2SAT() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	readLine := func() string {
		line, _ := reader.ReadString('\n')
		return strings.TrimSpace(line)
	}

	T, _ := strconv.Atoi(readLine())

	for t := 0; t < T; t++ {
		line := strings.Fields(readLine())
		n, _ := strconv.Atoi(line[0])
		k, _ := strconv.Atoi(line[1])

		N := 2 * n
		g := make([][]int, N)
		gr := make([][]int, N)

		for gi := 0; gi < k; gi++ {
			m, _ := strconv.Atoi(readLine())
			adj := make([]uint64, n)
			for i := 0; i < m; i++ {
				edge := strings.Fields(readLine())
				u, _ := strconv.Atoi(edge[0])
				v, _ := strconv.Atoi(edge[1])
				u--
				v--
				adj[u] |= 1 << v
				adj[v] |= 1 << u
			}

			for u := 0; u < n; u++ {
				for v := u + 1; v < n; v++ {
					bu := adj[u] &^ (1<<u | 1<<v)
					bv := adj[v] &^ (1<<u | 1<<v)

					if bu != bv {
						continue
					}

					has := (adj[u] & (1 << v)) != 0
					if has {
						a1 := varIndex(u, true)
						b1 := varIndex(v, false)
						g[a1] = append(g[a1], b1)
						gr[b1] = append(gr[b1], a1)

						a2 := varIndex(v, true)
						b2 := varIndex(u, false)
						g[a2] = append(g[a2], b2)
						gr[b2] = append(gr[b2], a2)
					} else {
						a1 := varIndex(u, false)
						b1 := varIndex(v, true)
						g[a1] = append(g[a1], b1)
						gr[b1] = append(gr[b1], a1)

						a2 := varIndex(v, false)
						b2 := varIndex(u, true)
						g[a2] = append(g[a2], b2)
						gr[b2] = append(gr[b2], a2)
					}
				}
			}
		}

		used := make([]bool, N)
		order := []int{}

		var dfs1 func(int)
		dfs1 = func(u int) {
			used[u] = true
			for _, v := range g[u] {
				if !used[v] {
					dfs1(v)
				}
			}
			order = append(order, u)
		}

		for i := 0; i < N; i++ {
			if !used[i] {
				dfs1(i)
			}
		}

		comp := make([]int, N)
		for i := range comp {
			comp[i] = -1
		}
		cid := 0

		var dfs2 func(int)
		dfs2 = func(u int) {
			comp[u] = cid
			for _, v := range gr[u] {
				if comp[v] == -1 {
					dfs2(v)
				}
			}
		}

		for i := N - 1; i >= 0; i-- {
			u := order[i]
			if comp[u] == -1 {
				dfs2(u)
				cid++
			}
		}

		ok := true
		for v := 0; v < n; v++ {
			if comp[varIndex(v, true)] == comp[varIndex(v, false)] {
				ok = false
				break
			}
		}

		if ok {
			fmt.Fprintln(writer, "Yes")
		} else {
			fmt.Fprintln(writer, "No")
		}
	}
}

func main() {
	solve2SAT()
}
